### PR TITLE
Monorepo vs microservices

### PR DESCRIPTION
**Pull request by dfellis**

### CODE REVIEW

File: src/monorepo-vs-microservices-is-an-abstraction-problem.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+-->
+
+We believe that all of the problems that causes one to migrate from a monorepo to microservices are due to problems in the abstractions of the languages used to write backend production code, that a new language can resolve these abstractions, and that a new language is *required* for some of these abstraction issues as the foundation of these languages does not fit with the reality they operate in.
+
+Minimizing latency when dealing with data stored in a cluster is the simplest problem to be abstracted and doable with current languages, but often very awkward and difficult to do. In a situation where a request comes in to node N1 but the relevant data is located on node N2, there are two possible choices to make:
```
Comment (depombo) [inline src/monorepo-vs-microservices-is-an-abstraction-problem.md:29]: Not sure how I feel about this being labeled as "Option 1" since most people use "Option 2" stateless compute calling into a distributed database that abstracts the concept of data spread across a cluster. I would switch the order maybe

File: src/monorepo-vs-microservices-is-an-abstraction-problem.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+1. Grab the relevant data from the request to N1 and push it and the code to execute to N2, then execute the code and push the result from N2 back to N1.
+2. Request the relevant data from N2 to N1, then execute the code using the relevant data from the request and return the result.
+
+Option 2 is taken the vast majority of the time because your distributed data is often in a distinct cluster (a database) from your request/response application layer, and doing the compute within the database is made difficult by needing to be written in another language from your main application language, requiring a context switch, and often made further difficult by this language primarily being a query language and being awkward or impossible to represent the computation desired.
```
Comment (depombo) [inline src/monorepo-vs-microservices-is-an-abstraction-problem.md:34]: Instead of "another language", I would specifically call out something SQL-like or a DB specific API which is what many NoSQL databases do

File: src/monorepo-vs-microservices-is-an-abstraction-problem.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+Even when the database supports the same language as your application, it often also has a differing set of standard libraries, making code reuse difficult if even possible. But option 1 is *often* the right choice when you have a complex computation you need to perform on a large amount of data that you can't just filter away, and so going through that pain when the execution time of option 2 is orders of magnitude more than option 1 is an optimization stable projects go through when traffic volume increases and the engineering cost to make that optimization becomes justifiable.
```
Comment (depombo) [inline src/monorepo-vs-microservices-is-an-abstraction-problem.md:36]: nit: maybe point out more specific examples of when people use the 2 different options?

File: src/monorepo-vs-microservices-is-an-abstraction-problem.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+Where `Tclosure` is the time to transmit the relevant metadata for the execution from N1 to N2, `Tresult` is the time to transmit the output result from N2 to N1, `Tdata` is the time to transmit the raw data from N2 to N1, and `Texecute` is the time required to perform the relevant computation, assumed to be identical on both sides here (but not the case when the execution has to be rewritten in a different language).
+
+Therefore, choose Option 1 when:
+
+```
+Tclosure + Tresult < Tdata
+```
+
+When the total data to compute on is less than the metadata from the request and the size out the output payload, then it makes sense to push the compute to the remote machine, otherwise it makes sense to pull the data from that node and perform the computation locally.
```
Comment (depombo) [inline src/monorepo-vs-microservices-is-an-abstraction-problem.md:53]: I like how this is laid out

File: src/monorepo-vs-microservices-is-an-abstraction-problem.md [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
...can produce a solution that will choose the correct execution location most of the time. The input data is from the closure and remote data pieces, and so the entropy, `E`, of the output result should not exceed that. Of course it is possible to create highly compressible data that still fits in that definition, but it is more likely that the compressibility of the input and output data are approximately equal, and that the output data has lower entropy than the input. Therefore, `Eresult < Edata + Eclosure` and almost always `Tresult < Tdata + Tclosure`. The range of possible entropies for the result is therefore between `[0..Edata + Eclosure]`. The exact distribution of entropies for `Eresult` depends on the use-case, but is likely a normal distribution with a low average and a long tail when considered in bulk across many different queries for many different remote executions. Therefore if we instead go with a conservative linear distribution, we overestimate the expected entropy of the output result somewhat and define `Eresult = (Edata + Eclosure) / 2`, and again assuming equal compressibility for equal entropy, we can substitute this back into our original inequality and get:
```
Comment (depombo) [inline src/monorepo-vs-microservices-is-an-abstraction-problem.md:59]: I would say node instead of side?

Comment (depombo) [reply src/monorepo-vs-microservices-is-an-abstraction-problem.md:59]: I feel like we can probably make this shorter without spelling out all assumptions and make the point more clear?

File: src/monorepo-vs-microservices-is-an-abstraction-problem.md [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+```
+Tclosure + (Tdata + Tclosure) / 2 < Tdata
+(3/2) * Tclosure < (1/2) * Tdata
+```
+
+This approximation will likely bias towards local execution when remote execution would have been the better bet, but generally when the stakes are low and the impact of one choice or the other has a low variance in the output latencies. And for static languages we can produce exact results in many circumstances (first by the type system, and later with a more intelligent compiler proving relations between `Eresult` and `Edata`). All of this analysis depends on being able to ship identically-behaving code between the nodes, and becomes considerably more difficult if the code in question must be rewritten in another language for the data storage side of things, hence why it has tended to be done on an adhoc, "squeaky wheel gets the grease" manner, and only when a developer has a "hunch" that it will pay off.
```
Comment (depombo) [inline src/monorepo-vs-microservices-is-an-abstraction-problem.md:66]: This could also be seen as more complex by folks, even if more performant because in the other "Option" developers don't have to think about cluster, remote/local execution, etc which should be maybe called out

File: src/monorepo-vs-microservices-is-an-abstraction-problem.md [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+What is different here is that `TexecutionUnit` doesn't fall out of the comparison as it did with "simple" remote execution. It is a vital component of deciding which level of parallelization should be taken, determining the smallest of `Tlocal`, `Tcores`, `ToneDatacenter` and `TallDatacenters` depends on the size and shape of the cluster as well as the size of the data, how finely it is diced up, and how long it takes to perform the transformation on any particular unit of the parallelizable data.
+
+And, as you may know, Turing complete languages cannot provide any guarantees on execution time; they may not even halt. [The Turing-Completeness Problem](./the-turing-completeness-problem.md) is a problem for most languages, but if you constrain the code that you are executing in parallel to a directed acyclic graph, then minimum and maximum execution times can be computed, and this can be used, with a similar averaging assumption, to make the parallelization decision decidable.
```
Comment (depombo) [inline src/monorepo-vs-microservices-is-an-abstraction-problem.md:101]: Turing complete code instead of languages?

File: src/monorepo-vs-microservices-is-an-abstraction-problem.md [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+That leaves the last two reasons that may be used in favor of microservices: vendoring different library versions and invalid code changes to on upgrades of parts of your cluster out of sync but the test suite not catching this. The former is simple: the language should simply adopt the dependency graph mechanism of dependency abstraction. The latter can also be simple: if you can have all of these "disparate" parts of your backend actually stay in the same codebase as the same singular deployment artifact, then it becomes impossible to get into such an invalid cluster state because you *do* basically Blue-Green deploy the entirety of production.
```
Comment (depombo) [inline src/monorepo-vs-microservices-is-an-abstraction-problem.md:107]: and the test suite not catching invalid code changes on upgrades to parts of your cluster.

?

**Review by depombo — COMMENTED** [review]
Looks good, mostly minor comments!

File: src/monorepo-vs-microservices-is-an-abstraction-problem.md [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+3. Latency-sensitive endpoints start impacting the architecture of the application layer. If the latency-sensitive piece is mostly read-bound, a cache layer is added either above the latency-sensitive endpoint, or between the application layer and the database layer, or both. If the latency-sensitive piece is write-bound, scaling up the database layer is often the first tack taken. If the latency-sensitive piece is both read-and-write bound in a kind of state machine way, either pushing this logic to the client-side or bolting on a second database (often NoSQL) dedicated to this small part of the application logic is done. This is the second point where breaking out into microservices may be debated, as the additional complexity in the request-handling for these endpoints can negatively impact development velocity on the "regular" endpoints as there may be extra "ceremony" to endpoint development that most of the developers do not care about and potentially don't fully understand, especially if homegrown with ill-defined terminology and separation-of-concerns.
```
[the remark this answers is not in the document]
Comment (dfellis) [reply src/monorepo-vs-microservices-is-an-abstraction-problem.md:9]: Tack was on purpose. It means a course correction on a boat to sail into the wind, I think, so I am talking about the first course correction on the architecture to handle the headwind of the problem at hand.

[the remark this answers is not in the document]
Comment (depombo) [reply src/monorepo-vs-microservices-is-an-abstraction-problem.md:9]: ah got it 👍

[the remark this answers is not in the document]
Comment (dfellis) [reply src/monorepo-vs-microservices-is-an-abstraction-problem.md:9]: Just confirming, my definition is right, and it's definition 4 here: https://www.merriam-webster.com/dictionary/tack

File: src/monorepo-vs-microservices-is-an-abstraction-problem.md [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
+At this point, it often devolves into a point and counterpoint between the two parties. The proponents of microservices arguing that separately deployable and testable code is already essentially microservices, but by being in the same repo, code changes that are impossible to migrate to but pass tests are possible. Eg, services A and B are both changed simultaneously to A' and B' and the test suite passes, but if you deploy service A -> A' first, it cannot interact with service B, and if you deploy B' first, it cannot interact with A, and since you can't deploy both services absolutely simultaneously, you're now going to cause a temporary outage. The proponents of monorepo argue that you can enforce that service mocks in the test suite cannot be changed at the same time code in the service is changed, so you can have the test suite enforce that the new code can talk to the old services and force the consideration of the migration path.
```
[the remark this answers is not in the document]
Comment (dfellis) [reply src/monorepo-vs-microservices-is-an-abstraction-problem.md:12]: The point of the paragraph was that both sides both start making arguments in favor of their own solution or against the alternative and they're *both right* and basically talking over each other. The exact examples I used here are less important.

Do you have any suggestions on how to make that clearer to the reader?

File: src/monorepo-vs-microservices-is-an-abstraction-problem.md [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+But is this actually the case? Let's take a look at the vendored library example more closely. Some of you saw that example and accepted it without question, while others of you were probably completely unfamiliar with it. For those of you coming from a C/C++ background, where each project must define their own way to vendor third party libraries because the build toolchain doesn't even consider the possibility, you likely saw this as a policy decision in the engineering org and the microservice proponents intentionally ignoring the alternative of allowing per-project vendored libraries within the monorepo. If you came from a Python or Java background, you likely saw this as entirely a technical issue, as dependencies are a list and you *can't* vendor multiple versions of the same library without jumping through hoops of `$PATH` munging or library name rewriting (and rewriting invocations in all other code using the version with the rewritten name). And if you came from a Node.js background you didn't understand it at all because your singular project probably has multiple versions of [ajv](https://www.npmjs.com/package/ajv) in your dependency tree and you don't even know *why*.
```
[the remark this answers is not in the document]
Comment (dfellis) [reply src/monorepo-vs-microservices-is-an-abstraction-problem.md:18]: I'm not sure what's being asked here?

File: src/monorepo-vs-microservices-is-an-abstraction-problem.md (hunk 1 above)
Comment (dfellis) [reply src/monorepo-vs-microservices-is-an-abstraction-problem.md:29]: Yeah, I can see that.

File: src/monorepo-vs-microservices-is-an-abstraction-problem.md (hunk 2 above)
Comment (dfellis) [reply src/monorepo-vs-microservices-is-an-abstraction-problem.md:34]: So I thought I did that in a later paragraph, but I can say it explicitly after "query language"?

Comment (dfellis) [reply src/monorepo-vs-microservices-is-an-abstraction-problem.md:34]:
```suggestion
Option 2 is taken the vast majority of the time because your distributed data is often in a distinct cluster (a database) from your request/response application layer, and doing the compute within the database is made difficult by needing to be written in another language from your main application language, requiring a context switch, and often made further difficult by this language primarily being a query language, likely SQL, and being awkward or impossible to represent the computation desired.
```

File: src/monorepo-vs-microservices-is-an-abstraction-problem.md (hunk 3 above)
Comment (dfellis) [reply src/monorepo-vs-microservices-is-an-abstraction-problem.md:36]: Do you have one in mind? I couldn't think of an example that wouldn't distract because there was too much setup explaining the problem and solution.

File: src/monorepo-vs-microservices-is-an-abstraction-problem.md (hunk 4 above)
Comment (dfellis) [reply src/monorepo-vs-microservices-is-an-abstraction-problem.md:53]: Yes, but I just realized I wrote `less` when I should have written `more` :man_facepalming:

Comment (dfellis) [reply src/monorepo-vs-microservices-is-an-abstraction-problem.md:53]:
```suggestion
Therefore, choose Option 1 when:

```
Tclosure + Tresult < Tdata
```

When the total data to compute on is more than the metadata from the request and the size out the output payload, then it makes sense to push the compute to the remote machine, otherwise it makes sense to pull the data from that node and perform the computation locally.
```

File: src/monorepo-vs-microservices-is-an-abstraction-problem.md (hunk 5 above)
Comment (dfellis) [reply src/monorepo-vs-microservices-is-an-abstraction-problem.md:59]: Yes, I was getting weary of how much I had to go into this, but it is also the more complex problem, so I'm not sure how much I can remove without oversimplifying things, as well.

File: src/monorepo-vs-microservices-is-an-abstraction-problem.md (hunk 6 above)
Comment (dfellis) [reply src/monorepo-vs-microservices-is-an-abstraction-problem.md:66]: I had hoped that mentioning developer experience, "hunches", rewriting the code, etc, as being necessary for it meant it was quite difficult to do right and not a good idea to do often and maintain code clarity.

Let me see how I can refactor this to make that clearer.

File: src/monorepo-vs-microservices-is-an-abstraction-problem.md (hunk 8 above)
Comment (dfellis) [reply src/monorepo-vs-microservices-is-an-abstraction-problem.md:107]: If you're doing a Blue-Green deploy of the entirety of production because it's simple and fast, then you're never actually upgrading parts of the cluster separate from the others?

Comment (dfellis) [reply src/monorepo-vs-microservices-is-an-abstraction-problem.md:107]: Any suggestions on how I can make that clearer?

File: src/monorepo-vs-microservices-is-an-abstraction-problem.md [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+
+There are other reasons why one would perform remote execution, though. If the data in question is too large to fit into the memory of a single node, splitting it across multiple nodes and performing all analysis of it via remote execution allows computation that would otherwise need to be awkwardly rewritten in a streaming style with disk backing (and potentially much slower, too). And in a computationally intensive problem, if the work can be parallelized, it *may* make sense to do so. There are multiple levels of parallelization decisions:
+
+1. Don't parallelize even if its possible.
+2. Parallelize across multiple CPU cores on the same node.
+3. Parallelize across multiple nodes in the same datacenter.
+4. Parallelize across multiple nodes in multiple datacenters.
+
+For the exact same parallelizable code, all four permutations are valid to consider, depending on the size of the payload an the execution time of the parallelizable unit.
```
Comment (cdmistman) [inline src/monorepo-vs-microservices-is-an-abstraction-problem.md:75]:
```suggestion
For the exact same parallelizable code, all four permutations are valid to consider, depending on the size of the payload and the execution time of the parallelizable unit.
```

**Review by cdmistman — COMMENTED** [review]
Looks great! Besides @depombo 's suggestions this is a good read. I'm a little wary of mentioning testing since there isn't a built-in testing framework so all Alan tests need to be tested through api endpoints and/or manually running the program

**Comment by cdmistman** [comment]
yeah, didn't mean to imply that a built-in testing framework would never come, just that it's not being worked on yet (besides the mock system you mentioned)

File: src/monorepo-vs-microservices-is-an-abstraction-problem.md [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+
+We believe that all of the problems that causes one to migrate from a monorepo to microservices are due to problems in the abstractions of the languages used to write backend production code, that a new language can resolve these abstractions, and that a new language is *required* for some of these abstraction issues as the foundation of these languages does not fit with the reality they operate in.
+
+Minimizing latency when dealing with data stored in a cluster is the simplest problem to be abstracted and doable with current languages, but often very awkward and difficult to do. In a situation where a request comes in to node N1 but the relevant data is located on node N2, there are two possible choices to make:
+
+1. Request the relevant data from N2 to N1, then execute the code using the relevant data from the request and return the result.
+2. Grab the relevant data from the request to N1 and push it and the code to execute to N2, then execute the code and push the result from N2 back to N1.
```
Comment (aguillenv) [inline src/monorepo-vs-microservices-is-an-abstraction-problem.md:27]: nit: I'd probably rephrase this as `Grab the relevant data from the request to N1 and push it along with the code to execute to N2, then execute the code on N2 and push the result back to N1.`

File: src/monorepo-vs-microservices-is-an-abstraction-problem.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (aguillenv) [reply src/monorepo-vs-microservices-is-an-abstraction-problem.md:12]: Agree, at first was a little confusing. Probably just adding a connector like `Meanwhile, the proponents of monorepo argue...` or `For the same case, the proponents of monorepo argue...` could make it clearer?

File: src/monorepo-vs-microservices-is-an-abstraction-problem.md (hunk 8 above)
Comment (aguillenv) [reply src/monorepo-vs-microservices-is-an-abstraction-problem.md:102]: I think it's ok as is.

**Review by aguillenv — APPROVED** [review]
